modules/security-jwt/src/main/java/org/opencastproject/security/jwt/JWTVerifier.java

```diff
@@ -21,55 +21,73 @@
 import org.springframework.expression.ExpressionParser;
-import org.springframework.expression.ParseException;
 import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.expression.spel.support.StandardEvaluationContext;
 import org.springframework.util.Assert;
 
+import java.util.Date;
 import java.util.List;
 
 /**
  * Helper class to verify JWTs.
  */
 public final class JWTVerifier {
 
+  /** Logging facility. */
+  private static final Logger logger = LoggerFactory.getLogger(JWTVerifier.class);
+
   private JWTVerifier() { }
 
   /**
    * Verifies a given JWT string with a given JWK provider and given claim constraints.
    *
    * @param token The JWT string.
-   * @param provider The JWK provider.
+   * @param retriever The JWK provider.
    * @param claimConstraints The claim constraints.
    * @return The decoded and verified JWT.
-   * @throws JwkException If the JWT cannot be verified successfully.
+   * @throws JOSEException If the JWT cannot be verified successfully.
+   * @throws java.text.ParseException If some part of the JWT cannot be parsed successfully.
    */
-  public static DecodedJWT verify(String token, GuavaCachedUrlJwkProvider provider, List<String> claimConstraints)
-          throws JwkException {
+  public static SignedJWT verify(String token, JWKSetProvider retriever, List<String> claimConstraints)
+          throws JOSEException, java.text.ParseException {
     Assert.notNull(token, "A token must be set");
-    Assert.notNull(provider, "A JWKS provider must be set");
-
-    DecodedJWT jwt = JWT.decode(token);
-
-    // First try with cache...
-    List<Algorithm> algorithms;
-    try {
-      algorithms = provider.getAlgorithms(jwt, false);
-      return verify(jwt, claimConstraints, algorithms);
-    } catch (JWTVerificationException | JwkException e) {
-      // ...then try again with forced fetch
-      // (recommended by e.g. https://openid.net/specs/openid-connect-core-1_0.html#RotateSigKeys)
-      algorithms = provider.getAlgorithms(jwt, true);
-      return verify(jwt, claimConstraints, algorithms);
+    Assert.notNull(retriever, "A JWKS retriever must be set");
+
+    SignedJWT jwt = SignedJWT.parse(token);
+    JWSAlgorithm alg = jwt.getHeader().getAlgorithm();
+
+    List<JWK> jwkSet = retriever.getAll();
+
+    if (alg.equals(JWSAlgorithm.RS256) || alg.equals(JWSAlgorithm.RS384) || alg.equals(JWSAlgorithm.RS512)) {
+      RSAKey rsaKey = jwkSet.get(0).toRSAKey();
+      return verify(jwt, claimConstraints, new RSASSAVerifier(rsaKey));
+    } else if (alg.equals(JWSAlgorithm.ES256) || alg.equals(JWSAlgorithm.ES256K) || alg.equals(JWSAlgorithm.ES384)
+        || alg.equals(JWSAlgorithm.ES512)) {
+      ECKey ecKey = jwkSet.get(0).toECKey();
+      return verify(jwt, claimConstraints, new ECDSAVerifier(ecKey));
+    } else if (alg.equals(JWSAlgorithm.EdDSA) || alg.equals(JWSAlgorithm.Ed25519)) {
+      JWK jwk = jwkSet.get(0).toPublicJWK();
```

Comment:
All this code only tries the first key of the JWKS? Am I reading this correctly? Why is that? Was that the previous behavior as well?

Reply:
@joneug do you know what the previous behavior was? If we only supported the first entry, IMO this would be a bug and we should support JWKS with multiple entries.

Reply:
In the current implementation, all keys are fetched from the provider.

https://github.com/opencast/opencast/blob/8046ae9b7fd3a67bc9581a308e3710c95cebcabd/modules/security-jwt/src/main/java/org/opencastproject/security/jwt/JWTVerifier.java#L63-L73

The verification logic then considers all of them:

https://github.com/opencast/opencast/blob/8046ae9b7fd3a67bc9581a308e3710c95cebcabd/modules/security-jwt/src/main/java/org/opencastproject/security/jwt/JWTVerifier.java#L126-L154

Reply:
Thanks for pointing that out, should be fixed now.